qiskit_experiments/visualization/plotters/base_plotter.py

```diff
@@ -0,0 +1,453 @@
+from qiskit_experiments.visualization.style import PlotStyle
+
+
+class BasePlotter(ABC):
+    """An abstract class for the serializable figure plotters.
+
+    A plotter takes data from an experiment analysis class or experiment and plots a given figure using a
+    drawing backend. Sub-classes define the kind of figure created and the expected data.
+
+    Data is split into series and figure data. Series data is grouped by series name (str). For
+    :class:`CurveAnalysis`, this is the model name for a curve fit. For series data associated with a
+    single series name and figure data, data-values are identified by a data-key (str). Different data
+    per series and figure must have a different data-key to avoid overwriting values. Experiment and
+    analysis results can be passed to the plotter so appropriate graphics can be drawn on the figure
+    canvas. Series data is added to the plotter using :meth:`set_series_data` whereas figure data is
+    added using :meth:`set_figure_data`. Series and figure data are retrieved using :meth:`data_for` and
+    :attr:`figure_data` respectively.
+
+    Series data contains values to be plotted on a canvas, such that the data can be grouped into subsets
+    identified by their series name. Series names can be thought of as legend labels for the plotted
+    data, and as curve names for a curve-fit. Figure data is not associated with a series or curve and is
+    instead only associated with the figure. Examples include analysis reports or other text that is
+    drawn onto the figure canvas.
+
+    # TODO: Add example usage and description of options and plot-options.
+    """
+
+    def __init__(self, drawer: BaseDrawer):
+        """Create a new plotter instance.
+
+        Args:
+            drawer: The drawer to use when creating the figure.
+        """
+        # Data to be plotted, such as scatter points, interpolated fits, and confidence intervals
+        self._series_data: Dict[str, Dict[str, Any]] = {}
+        # Data for figure-wide drawing, unrelated to series data, such as text or fit reports.
+        self._figure_data: Dict[str, Any] = {}
+
+        # Options for the plotter
+        self._options = self._default_options()
+        # Plotter options that have changed, for serialization.
+        self._set_options = set()
+
+        # Plot options that are updated in the drawer when `plotter.figure()` is called
+        self._plot_options = self._default_plot_options()
+        # Plot options that have changed, for serialization.
+        self._set_plot_options = set()
+
+        # The drawer backend to use for plotting.
+        self._drawer = drawer
+
+    @property
+    def drawer(self) -> BaseDrawer:
+        """The drawer being used by the plotter."""
+        return self._drawer
+
+    @drawer.setter
+    def drawer(self, new_drawer: BaseDrawer):
+        """Set the drawer to be used by the plotter."""
+        self._drawer = new_drawer
```

Comment:
If we can set and get this would it not make more sense to simply have `self.drawer = drawer` and remove the properties?

Reply:
My intention with this was to allow `BasePlotter` subclasses to execute code when `drawer` was changed. However, this is not used in this PR and I don't foresee it being used by any subclasses in the immediate future. I'll replace this with just a public `.drawer` variable.

Reply:
Addressed in 8035121.